src/cpp/StatisticsBackendData.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file StatisticsBackendData.cpp
+ */
+
+
+#include <map>
+#include <string>
+
+#include "StatisticsBackendData.hpp"
+#include "Monitor.hpp"
+
+#include <database/database.hpp>
+
+namespace eprosima {
+namespace statistics_backend {
+
+namespace database {
+
+class DatabaseEntityQueue;
+
+} // namespace database
+
+namespace details {
+
+
+void StatisticsBackendData::on_data_available(
+        EntityId domain_id,
+        EntityId entity_id,
+        DataKind data_kind)
+{
+    auto monitor = monitors_.find(domain_id);
+    assert(monitor != monitors_.end());
+    if (nullptr == monitor->second->domain_listener ||
+            !monitor->second->domain_callback_mask.is_set(CallbackKind::ON_DATA_AVAILABLE))
+    {
+        // No user listener, or mask deactivated
+        return;
+    }
+
+    if (!monitor->second->data_mask.is_set(data_kind))
+    {
+        // Data mask deactivated
+        return;
+    }
+
+    monitor->second->domain_listener->on_data_available(domain_id, entity_id, data_kind);
+}
+
+void StatisticsBackendData::on_domain_entity_discovery(
+        EntityId domain_id,
+        EntityId entity_id,
+        EntityKind entity_kind,
+        DiscoveryStatus discovery_status)
+{
+    auto monitor = monitors_.find(domain_id);
+    assert(monitor != monitors_.end());
+    if (monitor->second->domain_listener == nullptr)
+    {
+        // No user listener
+        return;
+    }
+
+    switch (entity_kind)
+    {
+        case EntityKind::PARTICIPANT:
+        {
+            if (!monitor->second->domain_callback_mask.is_set(CallbackKind::ON_PARTICIPANT_DISCOVERY))
+            {
+                // mask deactivated
+                return;
+            }
+
+            if (discovery_status == DISCOVERY)
+            {
+                monitor->second->participant_status_.on_instance_discovered();
+            }
+            else if (discovery_status == UNDISCOVERY)
+            {
+                monitor->second->participant_status_.on_instance_undiscovered();
+            }
+            monitor->second->domain_listener->on_participant_discovery(domain_id, entity_id,
```

Comment:
Should we avoid calling the listener when the status is neither `DISCOVERY` nor `UNDISCOVERY`?

Reply:
We decided to inform the application also in the case of modified QoS. The application can get the modified status from the QoS stored in the database, once it knows the modified entity ID.